lib/daru/category.rb

```diff
@@ -0,0 +1,447 @@
+        end
+      end.to_h
+    end
+
+    def min
+      assert_ordered 'min'
+
+      categories.first
+    end
+
+    def max
+      assert_ordered 'max'
+
+      categories.last
+    end
+
+    def sort!
+      assert_ordered 'sort'
+
+      # Build sorted index
+      old_index = @index.to_a
+      new_index = @cat_hash.values.map do |positions|
+        old_index.values_at(*positions)
+      end.flatten
+      @index = @index.class.new new_index
+
+      # Build sorted data
+      @cat_hash = categories.inject([{}, 0]) do |acc, cat|
+        hash, count = acc
+        cat_count = @cat_hash[cat].size
+        cat_count.times { |i| @array[count+i] = @map_int_cat.key(cat) }
+        hash[cat] = (count...(cat_count+count)).to_a
+        [hash, count + cat_count]
+      end.first
+
+      self
+    end
+
+    def contrast_code full=false
+      send("#{coding_scheme}_coding".to_sym, full)
+    end
+
+    def == other
+      size == other.size &&
+        to_a == other.to_a &&
+        index == other.index
+    end
+
+    def to_ints
+      @array
+    end
+
+    # Over rides original inspect for pretty printing in irb
+    def inspect spacing=20, threshold=15
+      row_headers = index.is_a?(MultiIndex) ? index.sparse_tuples : index.to_a
+
+      "#<#{self.class}(#{size})#{metadata && !metadata.empty? ? metadata.inspect : ''}>\n" +
+        Formatters::Table.format(
+          to_a.lazy.map { |v| [v] },
+          headers: @name && [@name],
+          row_headers: row_headers,
+          threshold: threshold,
+          spacing: spacing
+        )
+    end
+
+    # Convert to html for iruby
+    def to_html threshold=30
+      path =
+        if index.is_a?(MultiIndex)
+          File.expand_path('../iruby/templates/vector_mi.html.erb', __FILE__)
+        else
+          File.expand_path('../iruby/templates/vector.html.erb', __FILE__)
+        end
+      ERB.new(File.read(path).strip).result(binding)
+    end
+
+    def to_s
+      to_html
+    end
+
+    {
+      eq: :==,
+      not_eq: :!=,
+      lt: :<,
+      lteq: :<=,
+      mt: :>,
+      mteq: :>=
+    }.each do |method, operator|
+      define_method(method) do |other|
+        mod = Daru::Core::Query
+        if other.is_a?(Daru::Vector)
+          mod.apply_vector_operator operator, to_ints, other.to_ints
+        else
+          mod.apply_scalar_operator operator, @array, @map_int_cat.key(other)
+        end
+      end
+    end
+    alias :gt :mt
+    alias :gteq :mteq
+
+    def where bool_arry
+      Daru::Core::Query.vector_where to_a, @index.to_a, bool_arry, dtype, type
+    end
+
+    # TODO Cut function
+    def summary
+      values = {
+        size: size,
+        categories: categories.size,
+        max_freq: @cat_hash.values.map(&:size).max,
+        max_category: @cat_hash.keys.max_by { |cat| @cat_hash[cat].size },
+        min_freq: @cat_hash.values.map(&:size).min,
+        min_category: @cat_hash.keys.min_by { |cat| @cat_hash[cat].size }
+      }
+      
```

Comment:
I think that looks fine.

A related method that I find very useful, when doing interactive data analysis, is something like R's [table](http://stat.ethz.ch/R-manual/R-devel/library/base/html/table.html), which can display the count of each category.